### PR TITLE
Show winner - Added player class

### PR DESCRIPTION
**Pull request by JulesRoadknight**

### CODE REVIEW

File: lib/board.rb [hunk 1]
```diff
@@ -0,0 +1,136 @@
+class Board
+
+    attr_reader( :current_player, :winner, :board_middle, :turn_count, :max_turns, :board_size )
+    BOARD_ROW = '-'
+    BOARD_COLUMM = '|'
+    BEGINNING_AND_END_LENGTH = 2
+
+    def initialize(board_size = 3)
```
Comment (nickbdyer) [inline lib/board.rb]: This method feels very large. I think you are introducing unnecessary complexity by tracking all of these bits of data separately.

File: spec/player_spec.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+describe Player do
+    before(:each) do
+        @player = Player.new
+    end
+
+    it 'Starts with player O' do
+        expect(@player.current_player).to eq 'O'
+    end
+
+    # it 'Can select a move' do
```
Comment (nickbdyer) [inline spec/player_spec.rb]: Don't commit commented code, especially when it's a test.

File: spec/player_spec.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    #     column = (player_input - 1) % 3
+    #     @board.make_move(player_input, row, column)
+    #     expect(@player.square).to eq(1)
+    # end
+
+    it 'Can change to player X' do
+        expect { @player.change_player }.to output("X's turn\n").to_stdout
+    end
+
+    it 'Starts with player O' do
```
Comment (nickbdyer) [inline spec/player_spec.rb]: This looks like a duplicate of the test on line 8?

File: spec/features/can_win_diagonally_spec.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+require 'player'
```
Comment (nickbdyer) [inline spec/features/can_win_diagonally_spec.rb]: I'd be ok with all these tests being in the same test file. You are testing the same functionality albeit with different inputs.

File: lib/controller.rb [hunk 5]
```diff
@@ -0,0 +1,8 @@
+require_relative 'board'
```
Comment (nickbdyer) [inline lib/controller.rb]: As far as I can tell this class is only used from its corresponding spec file. Not sure why it exists in this PR?

File: player_test.rb [hunk 6]
```diff
@@ -0,0 +1,4 @@
+require './lib/player.rb'
```
Comment (nickbdyer) [inline player_test.rb]: As far as I can tell this file does nothing. The method it calls in Player doesn't exist?

File: lib/player.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
+require_relative 'board'
+
+class Player
```
Comment (nickbdyer) [inline lib/player.rb:1]: How many responsibilities does this class have?

Comment (nickbdyer) [reply lib/player.rb:1]: We can discuss this afternoon.

File: lib/board.rb [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    end
+
+    def check_column(moves_to_check)
+        for square in moves_to_check do
+            @winner = true if moves_to_check.include?(square + @board_size) && moves_to_check.include?(square + @board_size * 2)
+        end
+    end
+
+    def check_row(moves_to_check)
+        # for i in 0..@board_size do
```
Comment (nickbdyer) [inline lib/board.rb]: Please don't check in commented code.

File: lib/board.rb [hunk 9]
```diff
@@ -0,0 +1,136 @@
+class Board
```
Comment (nickbdyer) [inline lib/board.rb:1]: How many responsibilities does this class have?

File: lib/board.rb [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        @x_moves = []
+        @o_moves = []
+        @board_top_and_bottom = []
+        @board_middle = []
+        @board_middle_section = []
+        @board_size = board_size
+        @turn_count = 0
+        @max_turns = board_size * board_size
+        @winner = false
+        set_board
```
Comment (nickbdyer) [inline lib/board.rb]: Idiomatic indentation for Ruby should be 2 spaces I think.

**Review by nickbdyer — COMMENTED** [review]
image

We don't want your tests doing this. Can you think of a way to prevent it? 

Also, please enable `--format=documentation` (I think that's the flag) on your rspec tests. 

Final note, in enabling the above myself, I think it's become clear that `rspec` isn't actually running all your tests?

File: lib/board.rb [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+  def view_board
+    system('clear')
+    reset_board
+    assemble_board
+    puts @board
+  end
+
+  def check_square_is_free(row, column)
+    @board_middle[row][column] != 'X' && @board_middle[row][column] != 'O'
```
Comment (mercebc) [inline lib/board.rb:37]: Hardcoding the symbols of the players causes duplication of knowledge (DRY). Can you think of a way of only defining them in only one place?

File: lib/board.rb [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+
+  def board_top_and_bottom
+    board_top_and_bottom = []
+    (@board_size + BEGINNING_AND_END_LENGTH).times do
+      board_top_and_bottom.push(BOARD_ROW)
+    end
+    board_top_and_bottom = [board_top_and_bottom.join]
+  end
+
+  def assemble_board
```
Comment (mercebc) [inline lib/board.rb:60]: Assembling the board to be displayed feels like it’s not responsibility of the board itself. I would extract those functions into a different class

File: spec/features/can_win_on_other_board_sizes_spec.rb [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    game = Game.new
+    allow(game.input).to receive(:gets).and_return('1', '2', '6', '5', '11', '12', '21', '17', '16', '22')
+    game.go
+    expect(game.winner).to be true
+  end
+
+  it 'Win with column 7x7' do
+    allow_any_instance_of(Input).to receive(:gets).and_return('7')
+    game = Game.new
+    allow(game.input).to receive(:gets).and_return('1', '2', '9', '5', '17', '12', '25', '18', '33', '22', '41', '30', '49')
```
Comment (mercebc) [inline spec/features/can_win_on_other_board_sizes_spec.rb:23]: Tests that are easy to understand simplify the way we can detect the problem if the test fails. In this case, to improve readability of this test, you could create a private function with a name that describes what the function does?

**Pull request merged**